File: JeongSH/Programmers/DP/RoadToSchool.java
/*

    무지성 dfs 풀이법. 정확성 모두 통과 효율성 모두 시간초과
    이유: worst-case dfs 시간복잡도.
    모든 길을 전부 들어가서 answer 값을 ++ 하는 방식.

    시간복잡도:T(n) = O(V+E)

 */
public class RoadToSchool {
    static int DIV_VALUE = 1000000007;
    static int answer=0;
    public static void main(String args[]){
        int m=4;
        int n=3;
        int[][] puddles = {{2,2}};
        dfs(puddles,m,n,0,0);
        System.out.println(answer);
    }
    public static void dfs (int[][] puddle , int targetX, int targetY, int nowX,int nowY){
        if(nowX==targetX-1 && nowY==targetY-1){ answer=(answer+1)%DIV_VALUE;return;}    // 목적지에 도달한 경우이므로 answer+1

        if(nowX>= targetX || nowY >= targetY) {return;}                                 // 목적지를 넘어선 경우. 맵을 넘어감

        for(int i=0; i<puddle.length; i++){
            if(nowX == (puddle[i][0]-1) && nowY == (puddle[i][1]-1)) { return; }        // puddle을 만남.
        }

        dfs(puddle,targetX,targetY,nowX+1,nowY);                                 // 위의 경우가 아닌경우 x+1
        dfs(puddle,targetX,targetY,nowX,nowY+1);                                 // 위의 경우가 아닌경우 y+1
    }
}

File: 6loss0m/Programmers/DP/roadToSchool.py
def solution(m, n, puddles):
    answer = [[0]*(m+1) for _ in range(n+1)]
    answer[1][1] = 1 # 집이 있는 좌표
    for i in range(1, n+1):
        for j in range(1, m+1):
            # 집인 경우
            if i == 1 and j == 1:
                continue
            # 웅덩이 인 경우
            elif [j,i] in puddles:
                answer[i][j] = 0
            # 그외의 경우
            # 해당 지역으로 이동하는 경우는 위에서 오거나, 왼쪽에서 와야하기
            # 때문에 그 두가지의 경우의 수를 더해준다.
            else:
                answer[i][j] = answer[i][j-1] + answer[i-1][j]
    return answer[n][m] % 1000000007